File: requery-android/src/main/java/io/requery/android/sqlite/BaseConnection.java
/*
 * Copyright 2016 requery.io
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.requery.android.sqlite;

import java.sql.Array;
import java.sql.Blob;
import java.sql.CallableStatement;
import java.sql.Clob;
import java.sql.Connection;
import java.sql.NClob;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLClientInfoException;
import java.sql.SQLException;
import java.sql.SQLFeatureNotSupportedException;
import java.sql.SQLWarning;
import java.sql.SQLXML;
import java.sql.Savepoint;
import java.sql.Statement;
import java.sql.Struct;
import java.util.Map;
import java.util.Properties;

/**
 * Base {@link Connection} implementation.
 *
 * @author Nikhil Purushe
 */
public abstract class BaseConnection implements Connection {

    protected boolean autoCommit;
    protected int transactionIsolation;
    protected int holdability;
    protected Properties clientInfo;
    protected int savePointId;

    protected BaseConnection() {
        autoCommit = true;
        holdability = ResultSet.HOLD_CURSORS_OVER_COMMIT;
        clientInfo = new Properties();
        transactionIsolation = TRANSACTION_SERIALIZABLE;
    }

    protected abstract void ensureTransaction();

    protected abstract void execSQL(String sql) throws SQLException;

    @Override
    public void clearWarnings() throws SQLException {
    }

    @Override
    public void close() throws SQLException {
        // db must be closed outside of the connection
    }

    @Override
    public void commit() throws SQLException {

    }

    @Override
    public Statement createStatement(int resultSetType, int resultSetConcurrency)
            throws SQLException {
        return createStatement(resultSetType,
                resultSetConcurrency, ResultSet.HOLD_CURSORS_OVER_COMMIT);
    }

    @Override
    public boolean getAutoCommit() throws SQLException {
        return autoCommit;
    }

    @Override
    public String getCatalog() throws SQLException {
        return null;
    }

    @Override
    public int getHoldability() throws SQLException {
        return holdability;
    }

    @Override
    public int getTransactionIsolation() throws SQLException {
        return transactionIsolation;
    }

    @Override
    public Map<String, Class<?>> getTypeMap() throws SQLException {
        return null;
    }

    @Override
    public SQLWarning getWarnings() throws SQLException {
        return null;
    }

    @Override
    public String nativeSQL(String sql) throws SQLException {
        return sql;
    }

    @Override
    public CallableStatement prepareCall(String sql) throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public CallableStatement prepareCall(String sql, int resultSetType, int resultSetConcurrency)
            throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public CallableStatement prepareCall(String sql, int resultSetType, int resultSetConcurrency,
                                         int resultSetHoldability) throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public PreparedStatement prepareStatement(String sql) throws SQLException {
        return prepareStatement(sql, Statement.NO_GENERATED_KEYS);
    }

    @Override
    public PreparedStatement prepareStatement(String sql, int[] columnIndexes) throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public PreparedStatement prepareStatement(String sql, int resultSetType,
                                              int resultSetConcurrency) throws SQLException {
        return prepareStatement(sql, resultSetType, resultSetConcurrency,
                ResultSet.HOLD_CURSORS_OVER_COMMIT);
    }

    @Override
    public void releaseSavepoint(Savepoint savepoint) throws SQLException {
        execSQL("release savepoint " + savepoint.getSavepointName());
    }

    @Override
    public void rollback(Savepoint savepoint) throws SQLException {
        execSQL("rollback to savepoint " + savepoint.getSavepointName());
    }

    @Override
    public void setAutoCommit(boolean autoCommit) throws SQLException {
        this.autoCommit = autoCommit;
        ensureTransaction();
    }

    @Override
    public void setCatalog(String catalog) throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public void setHoldability(int holdability) throws SQLException {
        this.holdability = holdability;
    }

    @Override
    public void setReadOnly(boolean readOnly) throws SQLException {
        throw new SQLFeatureNotSupportedException("cannot change readonly mode after db opened");
    }

    @Override
    public Savepoint setSavepoint() throws SQLException {
        return setSavepoint(null);
    }

    @Override
    public Savepoint setSavepoint(String name) throws SQLException {
        savePointId++;
        if (name == null) {
            name = "sp" + String.valueOf(savePointId);
        }
        execSQL("savepoint " + name);
        return new DatabaseSavepoint(savePointId, name);
    }

    @Override
    public void setTransactionIsolation(int level) throws SQLException {
        switch (level) {
            case TRANSACTION_NONE:
            case TRANSACTION_SERIALIZABLE:
            case TRANSACTION_READ_COMMITTED:
                execSQL("PRAGMA read_uncommitted = false");
                transactionIsolation = level;
                break;
            case TRANSACTION_READ_UNCOMMITTED:
                execSQL("PRAGMA read_uncommitted = true");
                transactionIsolation = level;
                break;
            case TRANSACTION_REPEATABLE_READ:
                throw new SQLFeatureNotSupportedException();
            default:
                throw new SQLException("invalid isolation " + level);
        }
    }

    @Override
    public void setTypeMap(Map<String, Class<?>> map) throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public Clob createClob() throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public Blob createBlob() throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public NClob createNClob() throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public SQLXML createSQLXML() throws SQLException {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean isValid(int timeout) throws SQLException {
        return !isClosed();
    }

    @Override
    public void setClientInfo(String name, String value) throws SQLClientInfoException {
        clientInfo.setProperty(name, value);
    }

    @Override
    public void setClientInfo(Properties properties) throws SQLClientInfoException {
        if(properties != null) {
            this.clientInfo = properties;
        }
    }

    @Override
    public String getClientInfo(String name) throws SQLException {
        return clientInfo.getProperty(name);
    }

    @Override
    public Properties getClientInfo() throws SQLException {
        return clientInfo;
    }

    @Override
    public Array createArrayOf(String typeName, Object[] elements) throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public Struct createStruct(String typeName, Object[] attributes) throws SQLException {
        throw new SQLFeatureNotSupportedException();
    }

    @Override
    public <T> T unwrap(Class<T> iface) throws SQLException {
        return null;
    }

    @Override
    public boolean isWrapperFor(Class<?> iface) throws SQLException {
        return false;
    }

    private static class DatabaseSavepoint implements Savepoint {

        private final int id;
        private final String name;

        DatabaseSavepoint(int id, String name) {
            this.id = id;
            this.name = name;
        }

        @Override
        public int getSavepointId() throws SQLException {
            return id;
        }

        @Override
        public String getSavepointName() throws SQLException {
            return name;
        }
    }
}

File: requery-android/src/main/java/io/requery/android/sqlite/BaseConnection.kt
/*
 * Copyright 2018 requery.io
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.requery.android.sqlite

import android.database.sqlite.SQLiteAccessPermException
import android.database.sqlite.SQLiteCantOpenDatabaseException
import android.database.sqlite.SQLiteConstraintException
import android.database.sqlite.SQLiteDatabaseCorruptException

import java.sql.Array
import java.sql.Blob
import java.sql.CallableStatement
import java.sql.Clob
import java.sql.Connection
import java.sql.NClob
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.SQLFeatureNotSupportedException
import java.sql.SQLIntegrityConstraintViolationException
import java.sql.SQLNonTransientException
import java.sql.SQLWarning
import java.sql.SQLXML
import java.sql.Savepoint
import java.sql.Statement
import java.sql.Struct
import java.util.Properties

/**
 * Base [Connection] implementation.
 *
 * @author Nikhil Purushe
 */
abstract class BaseConnection protected constructor() : Connection {

    private var _autoCommit: Boolean = false
    private var _transactionIsolation: Int = 0
    private var _holdability: Int = 0
    private var _clientInfo: Properties? = null
    private var _savePointId: Int = 0

    init {
        _autoCommit = true
        _holdability = ResultSet.HOLD_CURSORS_OVER_COMMIT
        _clientInfo = Properties()
        _transactionIsolation = Connection.TRANSACTION_SERIALIZABLE
    }

    protected abstract fun ensureTransaction()

    abstract fun execSQL(sql: String)

    override fun clearWarnings() {}

    override fun close() {
        // db must be closed outside of the connection
    }

    override fun commit() {}

    @Throws(SQLException::class)
    override fun createStatement(resultSetType: Int, resultSetConcurrency: Int): Statement {
        return createStatement(resultSetType,
                resultSetConcurrency, ResultSet.HOLD_CURSORS_OVER_COMMIT)
    }

    override fun getAutoCommit(): Boolean {
        return _autoCommit
    }

    override fun getCatalog(): String? {
        return null
    }

    override fun getHoldability(): Int {
        return _holdability
    }

    override fun getTransactionIsolation(): Int {
        return _transactionIsolation
    }

    override fun getTypeMap(): Map<String, Class<*>>? {
        return null
    }

    override fun getWarnings(): SQLWarning? {
        return null
    }

    override fun nativeSQL(sql: String): String {
        return sql
    }

    @Throws(SQLException::class)
    override fun prepareCall(sql: String): CallableStatement {
        throw SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    override fun prepareCall(sql: String, resultSetType: Int, resultSetConcurrency: Int): CallableStatement {
        throw SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    override fun prepareCall(sql: String, resultSetType: Int, resultSetConcurrency: Int,
                             resultSetHoldability: Int): CallableStatement {
        throw SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    override fun prepareStatement(sql: String): PreparedStatement {
        return prepareStatement(sql, Statement.NO_GENERATED_KEYS)
    }

    @Throws(SQLException::class)
    override fun prepareStatement(sql: String, columnIndexes: IntArray): PreparedStatement {
        throw SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    override fun prepareStatement(sql: String, resultSetType: Int,
                                  resultSetConcurrency: Int): PreparedStatement {
        return prepareStatement(sql, resultSetType, resultSetConcurrency,
                ResultSet.HOLD_CURSORS_OVER_COMMIT)
    }

    @Throws(SQLException::class)
    override fun releaseSavepoint(savepoint: Savepoint) {
        execSQL("release savepoint " + savepoint.savepointName)
    }

    @Throws(SQLException::class)
    override fun rollback(savepoint: Savepoint) {
        execSQL("rollback to savepoint " + savepoint.savepointName)
    }

    override fun setAutoCommit(autoCommit: Boolean) {
        this._autoCommit = autoCommit
        ensureTransaction()
    }

    @Throws(SQLException::class)
    override fun setCatalog(catalog: String) {
        throw SQLFeatureNotSupportedException()
    }

    override fun setHoldability(holdability: Int) {
        this._holdability = holdability
    }

    @Throws(SQLException::class)
    override fun setReadOnly(readOnly: Boolean) {
        throw SQLFeatureNotSupportedException("cannot change readonly mode after db opened")
    }

    @Throws(SQLException::class)
    override fun setSavepoint(): Savepoint {
        return setSavepoint(null)
    }

    @Throws(SQLException::class)
    override fun setSavepoint(name: String?): Savepoint {
        var name = name
        _savePointId++
        if (name == null) {
            name = "sp" + _savePointId.toString()
        }
        execSQL("savepoint $name")
        return DatabaseSavepoint(_savePointId, name)
    }

    @Throws(SQLException::class)
    override fun setTransactionIsolation(level: Int) {
        _transactionIsolation = when (level) {
            Connection.TRANSACTION_NONE, Connection.TRANSACTION_SERIALIZABLE, Connection.TRANSACTION_READ_COMMITTED -> {
                execSQL("PRAGMA read_uncommitted = false")
                level
            }
            Connection.TRANSACTION_READ_UNCOMMITTED -> {
                execSQL("PRAGMA read_uncommitted = true")
                level
            }
            Connection.TRANSACTION_REPEATABLE_READ -> throw SQLFeatureNotSupportedException()
            else -> throw SQLException("invalid isolation $level")
        }
    }

    @Throws(SQLException::class)
    override fun setTypeMap(map: Map<String, Class<*>>) {
        throw SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    override fun createClob(): Clob {
        throw SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    override fun createBlob(): Blob {
        throw SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    override fun createNClob(): NClob {
        throw SQLFeatureNotSupportedException()
    }

    override fun createSQLXML(): SQLXML {
        throw UnsupportedOperationException()
    }

    @Throws(SQLException::class)
    override fun isValid(timeout: Int): Boolean {
        return !isClosed
    }

    override fun setClientInfo(name: String, value: String) {
        _clientInfo!!.setProperty(name, value)
    }

    override fun setClientInfo(properties: Properties?) {
        if (properties != null) {
            this._clientInfo = properties
        }
    }

    override fun getClientInfo(name: String): String {
        return _clientInfo!!.getProperty(name)
    }

    override fun getClientInfo(): Properties? {
        return _clientInfo
    }

    override fun createArrayOf(typeName: String?, elements: kotlin.Array<out Any>?): Array {
        throw SQLFeatureNotSupportedException()
    }

    override fun createStruct(typeName: String?, attributes: kotlin.Array<out Any>?): Struct {
        throw SQLFeatureNotSupportedException()
    }

    override fun <T> unwrap(iface: Class<T>): T? {
        return null
    }

    override fun isWrapperFor(iface: Class<*>): Boolean {
        return false
    }

    private class DatabaseSavepoint
    internal constructor(private val id: Int, private val name: String) : Savepoint {

        override fun getSavepointId(): Int {
            return id
        }

        override fun getSavepointName(): String {
            return name
        }
    }

    companion object {

        @Throws(SQLException::class)
        fun throwSQLException(exception: android.database.SQLException) {

            if (exception is SQLiteConstraintException) {
                throw SQLIntegrityConstraintViolationException(exception)

            } else if (exception is SQLiteCantOpenDatabaseException ||
                    exception is SQLiteDatabaseCorruptException ||
                    exception is SQLiteAccessPermException) {

                throw SQLNonTransientException(exception)
            }
            throw SQLException(exception)
        }
    }
}
